pkg/validate/streaming.go

```diff
@@ -0,0 +1,129 @@
+
+package validate
+
+import (
+	"bytes"
+	//"io"
+	"net/url"
+
+	"github.com/kubernetes-incubator/cri-tools/pkg/framework"
+	remote "k8s.io/apimachinery/pkg/util/remotecommand"
+	"k8s.io/client-go/rest"
+	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
+	internalapi "k8s.io/kubernetes/pkg/kubelet/api"
+	runtimeapi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	defaultStreamServer string = "http://127.0.0.1:10250"
+)
+
+var _ = framework.KubeDescribe("Streaming", func() {
+	f := framework.NewDefaultCRIFramework()
+
+	var rc internalapi.RuntimeService
+	var ic internalapi.ImageManagerService
+
+	BeforeEach(func() {
+		rc = f.CRIClient.CRIRuntimeClient
+		ic = f.CRIClient.CRIImageClient
+	})
+
+	Context("runtime should support streaming interfaces", func() {
+		var podID string
+		var podConfig *runtimeapi.PodSandboxConfig
+
+		BeforeEach(func() {
+			podID, podConfig = createPodSandboxForContainer(rc)
+		})
+
+		AfterEach(func() {
+			By("stop PodSandbox")
+			rc.StopPodSandbox(podID)
+			By("delete PodSandbox")
+			rc.RemovePodSandbox(podID)
+		})
+
+		It("runtime should support exec [Conformance]", func() {
+			By("test create a default container")
+			containerID := createDefaultContainer(rc, ic, podID, podConfig, "container-for-exec-test")
+
+			By("start container")
+			startContainer(rc, containerID)
+
+			req := createDefaultExec(rc, containerID)
+
+			By("check the output of exec")
+			checkExec(rc, req)
+		})
+	})
+})
+
+func createDefaultExec(c internalapi.RuntimeService, containerID string) string {
+	By("exec default command in container: " + containerID)
+	req := &runtimeapi.ExecRequest{
+		ContainerId: containerID,
+		Cmd:         []string{"echo", "hello"},
+	}
+
+	resp, err := c.Exec(req)
+	framework.ExpectNoError(err, "failed to exec in container %q", containerID)
+	framework.Logf("Get exec url: " + resp.Url)
+	return resp.Url
+}
+
+func checkExec(c internalapi.RuntimeService, req string) {
+	localOut := &bytes.Buffer{}
+	localErr := &bytes.Buffer{}
+
+	url := parseURL(c, req)
+	conf := &rest.Config{
+		Host: url.Host,
+	}
+	e, err := remotecommand.NewExecutor(conf, "POST", url)
+	framework.ExpectNoError(err, "failed to create executor for %q", req)
+
+	err = e.Stream(remotecommand.StreamOptions{
+		SupportedProtocols: []string{remote.StreamProtocolV1Name},
+		Stdin:              nil,
+		Stdout:             localOut,
+		Stderr:             localErr,
+		Tty:                false,
+	})
+	framework.ExpectNoError(err, "failed to open streamer for %q", req)
+
+	Expect(localOut.String()).To(Equal("hello\n"), "The stdout of exec should be hello")
+	Expect(localErr.String()).To(BeEmpty(), "The stderr of exec should be empty")
+	framework.Logf("Check exec url %q succeed", req)
+}
+
+func parseURL(c internalapi.RuntimeService, req string) *url.URL {
+	version := getVersion(c)
+	if version.RuntimeName == "docker" {
+		req = defaultStreamServer + req
```

Comment:
It's better to make this robust by URL itself, e.g. only concat if scheme and host is not specified in the URL.